processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java

```diff
@@ -0,0 +1,536 @@
+  private static final float DEFAULT_MAX_LOAD_FACTOR = 0.7f;
+
+  private final AggregatorFactory[] aggregatorFactories;
+
+  // Limit to apply to results.
+  // If limit > 0, track hash table entries in a binary heap with size of limit.
+  // If -1, no limit is applied, hash table entry offsets are tracked with an unordered list with no limit.
+  private int limit;
+
+  // Indicates if the sorting order has fields not in the grouping key, used when pushing down limit/sorting.
+  // In this case, grouping key comparisons need to also compare on aggregators.
+  // Additionally, results must be resorted by grouping key to allow results to merge correctly.
+  private boolean sortHasNonGroupingFields;
+
+  // Min-max heap, used for storing offsets when applying limits/sorting in the BufferGrouper
+  private ByteBufferMinMaxOffsetHeap offsetHeap;
+
+  // ByteBuffer slices used by the grouper
+  private ByteBuffer totalBuffer;
+  private ByteBuffer hashTableBuffer;
+  private ByteBuffer offsetHeapBuffer;
+
+  // Updates the heap index field for buckets, created passed to the heap when
+  // pushing down limit and the sort order includes aggregators
+  private BufferGrouperOffsetHeapIndexUpdater heapIndexUpdater;
+  private boolean initialized = false;
+
+  public LimitedBufferGrouper(
+      final Supplier<ByteBuffer> bufferSupplier,
+      final Grouper.KeySerde<KeyType> keySerde,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final int bufferGrouperMaxSize,
+      final float maxLoadFactor,
+      final int initialBuckets,
+      final int limit,
+      final boolean sortHasNonGroupingFields
+  )
+  {
+    super(bufferSupplier, keySerde, aggregatorFactories, bufferGrouperMaxSize);
+    this.maxLoadFactor = maxLoadFactor > 0 ? maxLoadFactor : DEFAULT_MAX_LOAD_FACTOR;
+    this.initialBuckets = initialBuckets > 0 ? Math.max(MIN_INITIAL_BUCKETS, initialBuckets) : DEFAULT_INITIAL_BUCKETS;
+    this.limit = limit;
+    this.sortHasNonGroupingFields = sortHasNonGroupingFields;
+
+    if (this.maxLoadFactor >= 1.0f) {
+      throw new IAE("Invalid maxLoadFactor[%f], must be < 1.0", maxLoadFactor);
+    }
+
+    int offset = HASH_SIZE + keySize;
+    this.aggregatorFactories = aggregatorFactories;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+
+    // For each bucket, store an extra field indicating the bucket's current index within the heap when
+    // pushing down limits
+    offset += Ints.BYTES;
+    this.bucketSize = offset;
+  }
+
+  @Override
+  public void init()
+  {
+    if (initialized) {
+      return;
+    }
+    this.totalBuffer = bufferSupplier.get();
+
+    LimitedBufferGrouper.validateBufferCapacity(
```

Comment:
can validateBufferCapacity be private and called directly instead of qualifying with classname? It just looks like a private utility method to me.